# 30 E2 Project

* ist1113931 David Vasques (36%)
  
* ist1113779 Leonor Machado (36%)
  
* ist1110824 João Costa (28%)

Prof. Diogo Ferreira and Prof. José Farinha

Lab Shift number: BD25610L-06; BD25610L-011

## Introdução

Após o diagrama Entidade-Associação para a base de dados “Zoo” ter sido apresentado ao cliente numa primeira entrega, seguiram-se várias iterações para alinhar o desenho da base de dados de acordo com os requisitos do cliente, tendo-se finalmente convergido no esquema SQL apresentado abaixo. 

Pretende-se agora implementar esta base de dados como parte de um sistema de informação que permita a gestão e análise da venda de bilhetes e da popularidade dos vários recintos e animais. A popularidade é um requisito novo, que advém de uma campanha do zoo: cada portador de *bilhete* tem direito a votar no seu *recinto* favorito, o que, na base de dados, incrementa *votos* do *recinto* em 1 e muda *votou* para TRUE no *bilhete*.

`Nota` Optou-se por um sistema de informação separado para a gestão dos funcionários do zoo, que não será abordado no presente trabalho.

## PART I – Base de Dados

#### 0. Criação da Base de Dados

Crie a base de dados `Zoo` no PostgreSQL e execute (nesta base de dados) os comandos para a implementação do respectivo esquema.

In [1]:
%load_ext sql
%config SqlMagic.displaycon = 0
%config SqlMagic.displaylimit = 100
%config SqlMagic.feedback = 0
%sql postgresql+psycopg://app:app@postgres/app --alias zoo

In [2]:
%%sql zoo
DROP MATERIALIZED VIEW IF EXISTS vendas_zoo CASCADE;
DROP TABLE IF EXISTS acesso CASCADE;
DROP TABLE IF EXISTS bilhete CASCADE;
DROP TABLE IF EXISTS venda CASCADE;
DROP TABLE IF EXISTS animal;
DROP TABLE IF EXISTS especie;
DROP TABLE IF EXISTS recinto;
DROP TABLE IF EXISTS zona;
DROP TYPE IF EXISTS cat;
DROP TYPE IF EXISTS cnt;

CREATE TYPE cat AS ENUM(
  'Aves',
  'Carnívoros',
  'Herbívoros',
  'Mamíferos Marinhos',
  'Primatas',
  'Repteis'
);

CREATE TYPE cnt AS ENUM(
  'África',
  'América',
  'Asia',
  'Austrália',
  'Europa'
);

CREATE TABLE zona (
  id_zona SERIAL PRIMARY KEY,
  categoria cat,
  continente cnt,
  preco NUMERIC(4, 2) NOT NULL,
  CONSTRAINT chave_zona UNIQUE (categoria, continente)
);

CREATE TABLE recinto (
  id_recinto SERIAL PRIMARY KEY,
  id_zona INTEGER NOT NULL REFERENCES zona,
  votos INTEGER
);

CREATE TABLE especie (
  nome_cientifico VARCHAR(100) PRIMARY KEY CHECK (nome_cientifico ~ '^[A-Z][a-z]+ [a-z]+$'),
  nome_comum VARCHAR(100) NOT NULL UNIQUE,
  categoria cat NOT NULL,
  continente cnt NOT NULL
);

CREATE TABLE animal (
  id_animal SERIAL PRIMARY KEY,
  nome VARCHAR(80) NOT NULL,
  nome_cientifico VARCHAR(100) NOT NULL REFERENCES especie,
  id_recinto INTEGER NOT NULL REFERENCES recinto,
  data_nasc DATE,
  CONSTRAINT chave_animal UNIQUE (nome, nome_cientifico)
);

CREATE TABLE venda (
  no_venda SERIAL PRIMARY KEY,
  data_hora TIMESTAMP NOT NULL,
  nif_cliente CHAR(9) CHECK (nif_cliente ~ '^[0-9]{9}$')
);

CREATE TABLE bilhete (
  bid SERIAL PRIMARY KEY,
  desconto NUMERIC(4, 2) NOT NULL DEFAULT 0.00,
  votou BOOLEAN NOT NULL DEFAULT FALSE,
  no_venda INTEGER NOT NULL REFERENCES venda
);

CREATE TABLE acesso (
  bid INTEGER NOT NULL REFERENCES bilhete,
  id_zona INTEGER NOT NULL REFERENCES zona,
  PRIMARY KEY (bid, id_zona)
);

++
||
++
++

Verifique que todas as tabelas foram criadas na base de dados com o seguinte comando:

In [3]:
%sqlcmd tables

Name
zona
recinto
especie
animal
venda
bilhete
acesso


#### 1. Restrições de Integridade [4 valores]

Recorrendo a `Triggers apenas quando estritamente necessário`, implemente na base de dados “Zoo” as seguintes restrições de integridade:


1. `RI-1` Em zona, a categoria e o continente não podem ser simultaneamente `NULL`
* uma zona é sempre dedicada a uma categoria de animais, a um continente, ou a ambos.

In [4]:
%%sql zoo

ALTER TABLE zona
ADD CONSTRAINT ri1_check_categoria_continente
CHECK (categoria IS NOT NULL OR continente IS NOT NULL);

++
||
++
++

2. `RI-2` Um animal tem de estar alojado num recinto que esteja numa zona compatível
* se a zona tiver categoria definida, tem de corresponder à categoria da espécie do animal;
* se tiver continente definido, tem de corresponder ao continente da espécie do animal.

In [5]:
%%sql zoo

CREATE OR REPLACE FUNCTION check_animal_zona_compat()
RETURNS TRIGGER AS $$
DECLARE
    esp_cat cat;
    esp_cont cnt;
    z_cat cat;
    z_cont cnt;
BEGIN
    SELECT categoria, continente INTO esp_cat, esp_cont
    FROM especie WHERE nome_cientifico = NEW.nome_cientifico;

    SELECT z.categoria, z.continente INTO z_cat, z_cont
    FROM recinto r 
    JOIN zona z ON r.id_zona = z.id_zona
    WHERE r.id_recinto = NEW.id_recinto;

    IF (z_cat IS NOT NULL AND z_cat != esp_cat) OR
        (z_cont IS NOT NULL AND z_cont != esp_cont) THEN 
        RAISE EXCEPTION 'RI-2: O recinto não pertence a uma zona compatível com a espécie do animal.';
    END IF;

    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS trg_ri2_animal_compat ON animal;
CREATE TRIGGER trg_ri2_animal_compat
BEFORE INSERT OR UPDATE ON animal
FOR EACH ROW EXECUTE FUNCTION check_animal_zona_compat();

++
||
++
++

`RI-3` Todos os animais de uma espécie têm de estar alojados em recinto(s) de uma mesma zona
* `Nota` No entanto, podem existir várias zonas compatíveis com a espécie.

In [6]:
%%sql zoo

CREATE OR REPLACE FUNCTION check_zona_especie()
RETURNS TRIGGER AS $$
DECLARE
    zona_novato INTEGER;
BEGIN
    SELECT id_zona INTO zona_novato
    FROM recinto
    WHERE id_recinto = NEW.id_recinto;

    IF EXISTS (
        SELECT 1
        FROM animal a
        JOIN recinto r ON a.id_recinto = r.id_recinto
        WHERE a.nome_cientifico = NEW.nome_cientifico
          AND r.id_zona != zona_novato
          AND a.id_animal IS DISTINCT FROM NEW.id_animal
    ) THEN
        RAISE EXCEPTION 'RI-3: Todos os animais da mesma espécie têm de estar alojados na mesma zona.';
    END IF;

    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS trg_ri3_same_zone ON animal;
CREATE TRIGGER trg_ri3_same_zone
BEFORE INSERT OR UPDATE ON animal
FOR EACH ROW
EXECUTE FUNCTION check_zona_especie();

++
||
++
++

`RI-4` Após uma transação de venda
* uma venda tem de incluir pelo menos um bilhete com acesso a pelo menos uma zona do zoo.

In [7]:
%%sql zoo

CREATE OR REPLACE FUNCTION check_venda_valida()
RETURNS TRIGGER AS $$
BEGIN
    IF NOT EXISTS (
        SELECT 1 
        FROM bilhete 
        WHERE no_venda = NEW.no_venda
    ) THEN
        RAISE EXCEPTION 'RI-4: A venda % é inválida pois não inclui nenhum bilhete.', NEW.no_venda;
    END IF;

    IF EXISTS (
        SELECT 1 
        FROM bilhete b
        WHERE b.no_venda = NEW.no_venda
        AND NOT EXISTS (
            SELECT 1 
            FROM acesso a 
            WHERE a.bid = b.bid
        )
    ) THEN
        RAISE EXCEPTION 'RI-4: A venda % é inválida pois contém bilhete(s) sem acesso a qualquer zona.', NEW.no_venda;
    END IF;
    
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS trg_ri4_venda_valida ON venda;
CREATE CONSTRAINT TRIGGER trg_ri4_venda_valida
AFTER INSERT ON venda
DEFERRABLE INITIALLY DEFERRED
FOR EACH ROW EXECUTE FUNCTION check_venda_valida();

++
||
++
++

#### 2. Preenchimento da Base de Dados [3 valores]

##### Critérios de Avaliação

* Cumprimento dos requisitos de cobertura  
* Resultado não vazio em todas as consultas do ponto 5

`Nota`: Caso não consiga implementar alguma(s) das restrições de integridade do ponto 1, deve ainda assim assegurar-se que o preenchimento da base dados as cumpre. 

Pode utilizar ferramentas de **IA generativo**, scripts ou qualquer outro meio para gerar os comandos INSERT. Preencha todas as tabelas da base de dados de forma consistente, após execução do ponto anterior para garantir que são respeitadas todas as restrições de integridade ali expressas, e com os seguintes **requisitos de cobertura** adicionais:

1. Têm de haver pelo menos 7 ***zonas***, das quais:
* O preço de acesso a cada zona deve estar entre 5€ e 30€.
* pelo menos 3 são dedicadas exclusivamente a uma *categoria* (sem *continente* preenchido);
* pelo menos 2 são dedicadas exclusivamente a um *continente* (sem *categoria* preenchida);
* pelo menos 2 têm *categoria* e *continente* preenchidos.

`Nota` As 2 ou mais zonas que têm *categoria* e *continente* preenchidos têm de partilhar entre todas elas um único valor de um desses dois atributos (sendo naturalmente o outro diferente, devido à restrição de chave candidata da tabela), e o valor desse atributo não pode ocorrer sozinho (i.e., qualquer zona que tenha o valor desse atributo não pode ter o outro atributo a NULL). O valor desse atributo corresponde à “especialidade” do zoo.  
* `Exemplo 1` Um zoo pode ter como especialidade herbívoros, e nesse caso terá de ter pelo menos 2 zonas com categoria “herbívoro” e continentes diferentes, não podendo ter nenhuma zona com categoria “herbívoro” sem continente definido. Terá adicionalmente pelo menos 3 zonas dedicadas a outras categorias que não “herbívoro” e 2 zonas dedicadas a continentes.  
* `Exemplo 2` Outro zoo pode ter como especialidade animais de África, e nesse caso terá de ter pelo menos 2 zonas com continente “África” e categorias diferentes, não podendo ter nenhuma zona com continente “África” sem categoria definida. Terá adicionalmente pelo menos 3 zonas dedicadas a categorias e 2 zonas dedicadas a continentes que não “África”.


In [8]:
%%sql zoo
    
INSERT INTO zona (categoria, continente, preco) VALUES
('Herbívoros', 'África', 10.00), -- 1. Especialidade (Ambos preenchidos)
('Herbívoros', 'Asia', 12.00),   -- 2. Especialidade (Ambos preenchidos)
('Carnívoros', NULL, 15.00),     -- 3. Só categoria
('Aves', NULL, 8.00),            -- 4. Só categoria
('Primatas', NULL, 10.00),       -- 5. Só categoria
(NULL, 'Europa', 5.00),          -- 6. Só continente
(NULL, 'Austrália', 14.00);      -- 7. Só continente

++
||
++
++

2. Cada *zona* tem de ter entre 10 e 30 ***recintos***, e cada recinto deve ter no mínimo 0.1% dos votos totais (ver 5. abaixo). Os votos devem refletir os *acessos* dos *bilhetes*:
* O somatório global dos *votos* de todos os ***recintos*** tem de ser igual ao número de ***bilhetes*** vendidos com *votou* TRUE; 
* O somatório dos *votos* de todos os ***recintos*** de uma ***zona*** tem de ser menor ou igual ao número de ***bilhetes*** vendidos com acesso a essa ***zona*** e *votou* TRUE.

`Nota`: Pode preencher os votos dos ***recintos*** já no INSERT destes, ou alternativamente após o preenchimento dos *bilhetes*, por meio de comandos de UPDATE.

In [9]:
%%sql zoo

DO $$
DECLARE 
    z_id INTEGER;
BEGIN
    FOR z_id IN 1..7 LOOP
        FOR i IN 1..15 LOOP
            INSERT INTO recinto (id_zona, votos) VALUES (z_id, 0);
        END LOOP;
    END LOOP;
END $$;

++
||
++
++

3. Têm de haver pelo menos 200 ***espécies*** reais cobrindo todas as *categorias* e todos os *continentes*.  

In [10]:
%%sql zoo
    
DO $$
DECLARE
    cats cat[] := ARRAY['Aves', 'Carnívoros', 'Herbívoros', 'Mamíferos Marinhos', 'Primatas', 'Repteis'];
    conts cnt[] := ARRAY['África', 'América', 'Asia', 'Austrália', 'Europa'];
    nome_c VARCHAR;
BEGIN
    FOR i IN 1..210 LOOP
        -- Gerar um nome científico inventado mas válido para a Regex (Ex: 'Genusx speciesya')
        nome_c := 'Genus' || chr(97 + (i % 26)) || ' species' || chr(97 + ((i / 26) % 26)) || chr(97 + (i % 26));
        
        -- Alinhar as primeiras 180 espécies com as nossas zonas para garantir compatibilidade!
        IF i <= 30 THEN 
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, 'Herbívoros', 'África');
        ELSIF i <= 60 THEN
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, 'Herbívoros', 'Asia');
        ELSIF i <= 90 THEN
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, 'Carnívoros', conts[(i%5)+1]);
        ELSIF i <= 120 THEN
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, 'Aves', conts[(i%5)+1]);
        ELSIF i <= 150 THEN
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, 'Primatas', conts[(i%5)+1]);
        ELSIF i <= 180 THEN
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, cats[(i%6)+1], 'Europa');
        ELSE
            INSERT INTO especie (nome_cientifico, nome_comum, categoria, continente) VALUES (nome_c, 'Especie '||i, cats[(i%6)+1], 'Austrália');
        END IF;
    END LOOP;
END $$;

++
||
++
++

4. O número de ***animais*** de cada *espécie* deve ser entre 1 e 12, com média entre 2 e 3 animais
* Um animal só pode estar num *recinto* sozinho se há 3 ou menos ***animais*** dessa *espécie*.
* Têm de haver no zoo alguns *recintos* com:
    * apenas um ***animal***;
    * vários ***animais*** de uma só *espécie*;
    * vários ***animais*** de várias *espécies*.

In [11]:
%%sql zoo
    
DO $$
DECLARE
    rec RECORD;
    num_animais INT;
    z_id INT; r_id INT;
    animal_count INT := 1;
    rnum INT := 1; -- Novo contador para substituir o row_number()
BEGIN
    FOR rec IN SELECT nome_cientifico FROM especie LOOP
        -- Atribuir a zona correta consoante a ordem de criação
        IF rnum <= 30 THEN z_id := 1; ELSIF rnum <= 60 THEN z_id := 2; ELSIF rnum <= 90 THEN z_id := 3;
        ELSIF rnum <= 120 THEN z_id := 4; ELSIF rnum <= 150 THEN z_id := 5; ELSIF rnum <= 180 THEN z_id := 6;
        ELSE z_id := 7; END IF;

        -- Escolher 1 a 4 animais (média ~2.5) e um recinto aleatório dentro da zona
        num_animais := floor(random() * 4 + 1)::int;
        r_id := (z_id - 1) * 15 + floor(random() * 15 + 1)::int;
        
        FOR i IN 1..num_animais LOOP
            INSERT INTO animal (nome, nome_cientifico, id_recinto, data_nasc)
            VALUES ('Animal ' || animal_count, rec.nome_cientifico, r_id, '2023-01-01');
            animal_count := animal_count + 1;
        END LOOP;
        
        rnum := rnum + 1; -- Incrementa o contador manualmente
    END LOOP;
END $$;

++
||
++
++

5. Têm de haver ***vendas*** de ***bilhetes*** para todos os dias de 2026 até 11 de Junho (i.e., `[2026-01-01 - 2026-06-11]`), tais que:
* Haja pelo menos 1000 ***bilhetes*** nos dias de semana;
* Haja pelo menos 4000 ***bilhetes*** nos fins de semana;
* Cerca de metade dos ***bilhetes*** por dia sejam para crianças ou séniores, tendo 50% de *desconto*, e os restantes não têm *desconto*;
* Pelo menos 2% dos ***bilhetes*** de cada dia tenham `Acesso Total` (i.e., ***acesso*** a todas as *zonas*);
* Todas as ***zonas*** estejam em pelo menos 10 ***bilhetes*** por dia;
* Cada ***bilhete*** tenha ***acesso*** a 3 ou mais ***zonas***, e cada ***zona*** tem de figurar em pelo menos 25% dos ***bilhetes***;
* Tem de haver ***bilhetes*** para todas as combinações de 3 ou mais ***zonas*** (mas não necessariamente todos os dias);
* Pelo menos 75% dos ***bilhetes*** têm *votou* a TRUE.
`Nota` A implementação da RI-4 no ponto anterior obriga a que a inserção nestas três tabelas seja encapsulada numa transação. 

In [12]:
#Comandos usados no terminal para povoar as tabelas de venda
 
#psql -h postgres -U app -d app
#\i povoar_vendas.sql


## PART II – Sistema de Informação e Aplicação

#### 3. Desenvolvimento de Aplicação [4 valores]

##### Critérios de Avaliação

* Funcionalidade dos *endpoints*
* Uso correto de transações
* Prevenção de injeção de SQL
* Uso correto de métodos e códigos de erro HTTP

`Nota` Todo o código da aplicação deve ser incluído na entrega do projeto.

`Nota` A aplicação deve ainda estar disponível *online* para demonstração durante a discussão oral, no ambiente de desenvolvimento Docker providenciado para a disciplina.

Crie um protótipo de RESTful *web service* para venda de bilhetes e votação nos recintos por acesso programático à base de dados ‘zoo’ através de uma API que devolve respostas em JSON, análoga à demonstrada no Lab10. A API deve implementar os seguintes *endpoints REST*:

| Endpoint | Descrição |
| ----- | ----- |
| /zona/\<zona\>/ | OUTPUT: lista de todos os ***recintos*** da \<zona\>, contendo o *número* do ***recinto*** e as ***espécies*** nele contidas, indicando, para cada ***espécie***, os *nomes científico* e *comum*, e o número de ***animais*** da ***espécie*** que estão no ***recinto***. |
| /recinto/\<recinto\>/voto/\<bilhete\>/ | Assinala o voto do \<bilhete\> no \<recinto\>, atualizando as tabelas ***bilhete*** (marcando *votou* TRUE) e ***recinto*** (incrementando os *votos*). Devolve mensagem de erro informativa e diferenciada (e não assinala o voto) se o \<bilhete\> já tinha *votou* \= TRUE ou se o \<recinto\> não está contido numa zona a que o \<bilhete\> tinha acesso. |
| /venda/ | Executa  uma venda de um ou mais bilhetes, ***populando*** as tabelas ***venda***, ***bilhete*** e ***acesso***. INPUT: NIF do cliente \[opcional\] mais lista de bilhetes, em que cada bilhete consiste numa lista de zonas de acesso, e um desconto \[opcional\]. OUTPUT: O preço total da venda, e a lista dos bilhetes da venda com o número e preço de cada um.  |

A solução deve prezar pela segurança, **prevenindo** ataques por **injeção de** **SQL**, e garantindo que as **operações de escrita** estão encapsuladas em **transações** que cumprem os princípios ACID.

0. Itemize quais as estratégias que utilizou:

* Estratégia 1: Uso de parametrização de queries (%(...)s) em vez de concatenação de strings para prevenir injeção de SQL.
* Estratégia 2: Uso de contextos de transação (with conn.transaction():) para permitir o funcionamento dos triggers de integridade.

1. Copie para a célula abaixo a função do app.py zona_index

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/zona/<id_zona>/", methods=("GET",))
def zona_index(id_zona):
    """Mostra a lista de recintos de uma zona e as espécies/animais neles contidos."""

    with pool.connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """
                SELECT 
                    r.id_recinto,
                    e.nome_cientifico,
                    e.nome_comum,
                    COUNT(a.id_animal) AS num_animais
                FROM recinto r
                LEFT JOIN animal a ON r.id_recinto = a.id_recinto
                LEFT JOIN especie e ON a.nome_cientifico = e.nome_cientifico
                WHERE r.id_zona = %(id_zona)s
                GROUP BY r.id_recinto, e.nome_cientifico, e.nome_comum
                ORDER BY r.id_recinto;
                """,
                {"id_zona": id_zona},
            )
            rows = cur.fetchall()
            log.debug(f"Recuperadas {cur.rowcount} linhas para a zona {id_zona}.")

    recintos_dict = {}
    for row in rows:
        id_recinto = row[0]
        nome_cientifico = row[1]
        nome_comum = row[2]
        num_animais = row[3]

        if id_recinto not in recintos_dict:
            recintos_dict[id_recinto] = {
                "id_recinto": id_recinto,
                "especies": []
            }

        if nome_cientifico:
            recintos_dict[id_recinto]["especies"].append({
                "nome_cientifico": nome_cientifico,
                "nome_comum": nome_comum,
                "num_animais": num_animais
            })

    return jsonify(list(recintos_dict.values())), 200
``` 

2. Copie para a célula abaixo a função do app.py recinto_voto_save

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/recinto/<id_recinto>/voto/<bid>/", methods=("POST", "PUT"))
def recinto_voto_save(id_recinto, bid):
    """Regista o voto de um bilhete num determinado recinto."""

    with pool.connection() as conn:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT votou FROM bilhete WHERE bid = %(bid)s;",
                {"bid": bid}
            )
            bilhete = cur.fetchone()
            if not bilhete:
                return jsonify({"message": "Bilhete não encontrado.", "status": "error"}), 404
            
            if bilhete[0]: 
                return jsonify({"message": "Este bilhete já foi utilizado para votar.", "status": "error"}), 400

            cur.execute(
                "SELECT id_zona FROM recinto WHERE id_recinto = %(id_recinto)s;",
                {"id_recinto": id_recinto}
            )
            recinto = cur.fetchone()
            if not recinto:
                return jsonify({"message": "Recinto não encontrado.", "status": "error"}), 404
            
            id_zona_recinto = recinto[0]

            cur.execute(
                "SELECT 1 FROM acesso WHERE bid = %(bid)s AND id_zona = %(id_zona)s;",
                {"bid": bid, "id_zona": id_zona_recinto}
            )
            tem_acesso = cur.fetchone()
            if not tem_acesso:
                return jsonify({"message": "O bilhete não tem acesso à zona deste recinto.", "status": "error"}), 400

            try:
                with conn.transaction():
                    cur.execute(
                        "UPDATE bilhete SET votou = TRUE WHERE bid = %(bid)s;",
                        {"bid": bid}
                    )
                    cur.execute(
                        "UPDATE recinto SET votos = COALESCE(votos, 0) + 1 WHERE id_recinto = %(id_recinto)s;",
                        {"id_recinto": id_recinto}
                    )
                log.debug(f"Voto registado com sucesso. Bilhete {bid} -> Recinto {id_recinto}.")
            except Exception as e:
                return jsonify({"message": f"Erro interno ao processar transação: {str(e)}", "status": "error"}), 500

    return jsonify({"message": "Voto assinalado com sucesso.", "status": "success"}), 200
```

3. Copie para a célula abaixo a função do app.py venda_save

`Nota` Formate o seu código previamente e mantenha o bloco ```python para colorizar a sintaxe

```python
@app.route("/venda/", methods=("POST",))
def venda_save():
    """Executa a transação de venda de um ou mais bilhetes com os respetivos acessos."""
    
    payload = request.get_json()
    if not payload:
        return jsonify({"message": "Dados da requisição em falta.", "status": "error"}), 400

    nif_cliente = payload.get("nif_cliente")
    bilhetes_input = payload.get("bilhetes")

    if not bilhetes_input or not isinstance(bilhetes_input, list):
        return jsonify({"message": "A lista de bilhetes é obrigatória.", "status": "error"}), 400

    preco_total_venda = 0.0
    bilhetes_emitidos = []

    with pool.connection() as conn:
        with conn.cursor() as cur:
            try:
                with conn.transaction():
                    cur.execute(
                        """
                        INSERT INTO venda (data_hora, nif_cliente)
                        VALUES (NOW(), %(nif_cliente)s)
                        RETURNING no_venda;
                        """,
                        {"nif_cliente": nif_cliente}
                    )
                    no_venda = cur.fetchone()[0]

                    for idx, b_data in enumerate(bilhetes_input):
                        desconto = float(b_data.get("desconto", 0.00))
                        zonas = b_data.get("zonas", [])

                        if not zonas:
                            raise Exception(f"O bilhete no índice {idx} necessita de pelo menos uma zona de acesso.")

                        cur.execute(
                            """
                            INSERT INTO bilhete (desconto, votou, no_venda)
                            VALUES (%(desconto)s, FALSE, %(no_venda)s)
                            RETURNING bid;
                            """,
                            {"desconto": desconto, "no_venda": no_venda}
                        )
                        bid = cur.fetchone()[0]

                        preco_bilhete = 0.0
                        for id_zona in zonas:
                            cur.execute(
                                """
                                INSERT INTO acesso (bid, id_zona)
                                VALUES (%(bid)s, %(id_zona)s);
                                """,
                                {"bid": bid, "id_zona": id_zona}
                            )

                            cur.execute("SELECT preco FROM zona WHERE id_zona = %(id_zona)s;", {"id_zona": id_zona})
                            zona_res = cur.fetchone()
                            if not zona_res:
                                raise Exception(f"A zona com ID {id_zona} introduzida não existe.")

                            preco_base_zona = float(zona_res[0])
                            preco_bilhete += preco_base_zona * (1.0 - desconto)

                        preco_total_venda += preco_bilhete
                        bilhetes_emitidos.append({
                            "bilhete": bid,
                            "preco": round(preco_bilhete, 2)
                        })

                log.debug(f"Venda {no_venda} concluída com sucesso.")
            except Exception as e:
                return jsonify({"message": f"Falha na venda (Rollback executado): {str(e)}", "status": "error"}), 400

    output_venda = {
        "preco_total": round(preco_total_venda, 2),
        "bilhetes": bilhetes_emitidos
    }
    return jsonify(output_venda), 201

```

## PART III – Análise de Dados

A resolução das alíneas que se seguem **tem de obedecer às seguintes restrições**:
- Só pode haver **um único comando exterior** (CREATE, ALTER, UPDATE ou SELECT) e, no caso de envolver o comando SELECT, **não pode haver mais do que 1 nível de encadeamento de SELECTs**.
- **Não pode** recorrer aos operadores LIMIT ou WITH, ou a qualquer operador não coberto nos materiais da disciplina.


#### 4. Engenharia de Dados [2 valores]

##### Critérios de Avaliação

* Correção das soluções
* Eficiência das soluções
* Simplicidade das soluções

1. Crie uma vista materializada para auxiliar a direção do zoo a analisar as vendas de bilhetes e receitas do zoo e suas zonas, com o seguinte esquema:

    *vendas_zoo(bid, id_zona, mes, dia_da_semana, data, receita)*

    Em que teremos, para cada bilhete (*bid*) e zona (*id_zona*) a que o bilhete teve acesso, a data de venda do bilhete (e atributos derivados mes e dia_da_semana) e a receita da zona com o bilhete (i.e., o preço base da zona modificado pelo desconto).

In [26]:
%%sql zoo

DROP MATERIALIZED VIEW IF EXISTS vendas_zoo;
CREATE MATERIALIZED VIEW vendas_zoo AS
SELECT
    b.bid,
    a.id_zona,
    EXTRACT(MONTH FROM v.data_hora) AS mes,
    EXTRACT(ISODOW FROM v.data_hora) AS dia_da_semana,
    CAST(v.data_hora AS DATE) AS data,
    (z.preco * (1 - b.desconto)) AS receita
FROM venda v
JOIN bilhete b ON v.no_venda = b.no_venda
JOIN acesso a ON b.bid = a.bid
JOIN zona z ON a.id_zona = z.id_zona;

++
||
++
++

2. Modifique a tabela recinto adicionando um campo rentabilidade de tipo REAL

In [27]:
%%sql zoo
    
ALTER TABLE recinto
ADD COLUMN IF NOT EXISTS rentabilidade REAL;

++
||
++
++

3. Actualize a tabela recinto com o valor da rentabilidade sendo obtido pela receita total da zona onde está o recinto multiplicada pela fração entre os votos do recinto e o total de votos na zona. Pode recorrer à vista votos_zoo para realizar esta actualização.

In [28]:
%%sql zoo

UPDATE recinto r1
SET rentabilidade = COALESCE(
    (SELECT SUM(receita) FROM vendas_zoo vz WHERE vz.id_zona = r1.id_zona) *
    (r1.votos::REAL / NULLIF((SELECT SUM(votos) FROM recinto r2 WHERE r2.id_zona = r1.id_zona), 0)),0);

++
||
++
++

#### 5. Consultas [4 valores]

##### Critérios de Avaliação

* Correção das soluções
* Eficiência das soluções
* Simplicidade das soluções

Apresente a consulta SQL mais sucinta para cada um dos seguintes objetivos analíticos da empresa. Deve usar a vista vendas_zoo (exclusivamente ou em adição às outras tabelas) sempre que isso simplificar a consulta. Pode também usar operadores OLAP onde for adequado ao pedido.

1. Determinar o recinto mais rentável para cada zona, incluindo o valor da sua rentabilidade.

In [29]:
%%sql zoo

SELECT id_zona, id_recinto, rentabilidade
FROM recinto r1
WHERE rentabilidade = (
    SELECT MAX(rentabilidade)
    FROM recinto r2
    WHERE r1.id_zona = r2.id_zona
);

id_zona,id_recinto,rentabilidade
6,76,87738.52
1,4,175562.62
2,29,177767.92
3,31,250866.17
4,48,118429.64
5,62,177643.11
7,96,243734.6


2. Determinar se a aposta do zoo na sua especialidade está a compensar em termos de receitas, bilhetes ou votos, calculando as médias por zona, entre zonas da especialidade e zonas que não são da especialidade, de receitas globais, de bilhetes vendidos, e de votos recebidos.

In [30]:
%%sql zoo

SELECT
    (z.categoria IS NOT NULL AND z.continente IS NOT NULL) AS especialidade,
    AVG(vz_ag.receita_zona) AS media_receitas_por_zona,
    AVG(vz_ag.bilhetes_zona) AS media_bilhetes_por_zona,
    AVG(r_ag.votos_zona) AS media_votos_por_zona
FROM zona z
LEFT JOIN(
    SELECT id_zona, SUM(receita) AS receita_zona, COUNT(DISTINCT bid) AS bilhetes_zona
    FROM vendas_zoo
    GROUP BY id_zona
) vz_ag ON z.id_zona = vz_ag.id_zona
LEFT JOIN (
    SELECT id_zona, SUM(votos) AS votos_zona
    FROM recinto
    GROUP BY id_zona
) r_ag ON z.id_zona = r_ag.id_zona
GROUP BY (z.categoria IS NOT NULL AND z.continente IS NOT NULL);

especialidade,media_receitas_por_zona,media_bilhetes_por_zona,media_votos_por_zona
False,1367392.200000000000,175264.200000000000,32375.000000000000
True,1445775.500000000000,175288.500000000000,33064.500000000000


3. Determinar a percentagem de bilhetes com acesso a todas as zonas, globalmente e com *drill downs* independentes por dia da semana e por mês.

In [31]:
%%sql zoo

SELECT
    mes,
    dia_da_semana,
    ROUND(100.0 * COUNT(bid) FILTER (WHERE num_zonas = (SELECT COUNT(*) FROM zona)) / NULLIF(COUNT(bid), 0), 2) AS percentagem_acesso_total
FROM (
    SELECT bid, mes, dia_da_semana, COUNT(id_zona) AS num_zonas
    FROM vendas_zoo
    GROUP BY bid, mes, dia_da_semana
) bz
GROUP BY GROUPING SETS ((), (dia_da_semana), (mes));

mes,dia_da_semana,percentagem_acesso_total
None,None,2.99
None,3,2.87
None,1,2.81
None,4,2.91
None,6,3.04
None,7,3.03
None,2,3.00
None,5,3.03
3,None,2.97
1,None,3.06


4. Determinar que zonas podem ter um aumento no preço e que zonas devem ter uma redução no preço, analisando a média diária de bilhetes vendidos, globalmente e com vértices nas dimensões zona e mês.

In [32]:
%%sql zoo

SELECT
    id_zona,
    mes,
    ROUND((COUNT(DISTINCT bid)::NUMERIC / NULLIF(COUNT(DISTINCT data), 0)), 2) AS media_diaria_bilhetes
FROM vendas_zoo
GROUP BY CUBE (id_zona, mes)
ORDER BY id_zona, mes;

id_zona,mes,media_diaria_bilhetes
1,1,1094.90
1,2,1085.25
1,3,1092.81
1,4,1051.07
1,5,1150.55
1,6,902.18
1,None,1082.28
2,1,1093.16
2,2,1084.79
2,3,1092.48


#### 6. Índices [3 valores]

##### Critérios de Avaliação

* Utilidade dos índices
* Adequação  da justificação

É expectável que seja necessário executar consultas semelhantes **às consultas 1 e 2 do ponto anterior** diversas vezes ao longo do tempo, pelo que pretendemos otimizar o desempenho dessas consultas por meio da criação de índices. Crie sobre a vista vendas\_zoo ou sobre as restantes tabelas da base de dados o(s) índice(s) que achar mais indicados para fazer essa otimização, justificando a sua escolha com **argumentos teóricos** e com **demonstração prática** do ganho em eficiência do índice por meio do comando EXPLAIN ANALYSE. Deve procurar uma otimização coletiva das duas consultas, evitando criar índices excessivos, particularmente se estes trazem apenas ganhos incrementais. Nomeadamente, devido a preocupações com o armazenamento, não se considera vantajoso atingir planos de execução com INDEX ONLY SCAN se isso obriga a colocar no índice mais atributos do que estritamente necessário para conseguir um INDEX SCAN.

##### Consulta 1

1. Demonstração do custo (com EXPLAIN ANALYSE)

In [20]:
%%sql zoo
EXPLAIN ANALYSE
SELECT id_zona, id_recinto, rentabilidade
FROM recinto r1
WHERE rentabilidade = (
    SELECT MAX(rentabilidade)
    FROM recinto r2
    WHERE r1.id_zona = r2.id_zona
);

QUERY PLAN
Seq Scan on recinto r1 (cost=0.00..61374.50 rows=9 width=12) (actual time=0.038..0.747 rows=105.00 loops=1)
Filter: (rentabilidade = (SubPlan 1))
Buffers: shared hit=212
SubPlan 1
-> Aggregate (cost=33.15..33.16 rows=1 width=4) (actual time=0.006..0.006 rows=1.00 loops=105)
Buffers: shared hit=210
-> Seq Scan on recinto r2 (cost=0.00..33.12 rows=9 width=4) (actual time=0.002..0.005 rows=15.00 loops=105)
Filter: (r1.id_zona = id_zona)
Rows Removed by Filter: 90
Buffers: shared hit=210


2. Comandos de criação do(s) índice(s)

In [21]:
%%sql zoo
CREATE INDEX idx_recinto_zona_rent ON recinto(id_zona, rentabilidade DESC);

++
||
++
++

3. Demonstração do custo final (com EXPLAIN ANALYSE)

In [22]:
%%sql zoo
EXPLAIN ANALYSE
SELECT id_zona, id_recinto, rentabilidade
FROM recinto r1
WHERE rentabilidade = (
    SELECT MAX(rentabilidade)
    FROM recinto r2
    WHERE r1.id_zona = r2.id_zona
);

QUERY PLAN
Seq Scan on recinto r1 (cost=0.00..352.44 rows=1 width=12) (actual time=0.020..0.569 rows=105.00 loops=1)
Filter: (rentabilidade = (SubPlan 1))
Buffers: shared hit=212
SubPlan 1
-> Aggregate (cost=3.31..3.32 rows=1 width=4) (actual time=0.005..0.005 rows=1.00 loops=105)
Buffers: shared hit=210
-> Seq Scan on recinto r2 (cost=0.00..3.31 rows=1 width=4) (actual time=0.002..0.004 rows=15.00 loops=105)
Filter: (r1.id_zona = id_zona)
Rows Removed by Filter: 90
Buffers: shared hit=210


4. Justificação

Propusemos o índice composto idx_recinto_zona_rent ON recinto(id_zona, rentabilidade DESC) porque a consulta procura, para cada zona, o recinto com maior rentabilidade. Com este índice, torna-se possível encontrar o valor máximo por id_zona de forma eficiente, estruturando os dados para evitar a necessidade de ordenar ou verificar metodicamente toda a tabela, o que otimiza o tempo de execução e o custo de I/O. Como trade-off, o índice acarreta um custo associado ao espaço adicional de armazenamento e um impacto marginal nas operações de escrita.

Nota de análise do EXPLAIN: Sendo a tabela recinto muito pequena no atual ambiente de testes (apenas 105 registos), o Query Planner optou por manter um Seq Scan por ser estatisticamente mais rápido do que carregar as páginas do índice e incorrer no overhead de travessia da árvore B-Tree. Num cenário real de produção, com uma escala de milhares ou milhões de recintos, o Index Scan seria acionado pelo otimizador do PostgreSQL, fazendo com que a redução de I/O e a eficiência temporal se notassem de forma muito substancial.

##### Consulta 2

1. Demonstração do custo inicial (com EXPLAIN ANALYSE)

In [23]:
%%sql zoo
EXPLAIN ANALYSE
SELECT
    (z.categoria IS NOT NULL AND z.continente IS NOT NULL) AS especialidade,
    AVG(vz_ag.receita_zona) AS media_receitas_por_zona,
    AVG(vz_ag.bilhetes_zona) AS media_bilhetes_por_zona,
    AVG(r_ag.votos_zona) AS media_votos_por_zona
FROM zona z
LEFT JOIN(
    SELECT id_zona, SUM(receita) AS receita_zona, COUNT(DISTINCT bid) AS bilhetes_zona
    FROM vendas_zoo
    GROUP BY id_zona
) vz_ag ON z.id_zona = vz_ag.id_zona
LEFT JOIN (
    SELECT id_zona, SUM(votos) AS votos_zona
    FROM recinto
    GROUP BY id_zona
) r_ag ON z.id_zona = r_ag.id_zona
GROUP BY (z.categoria IS NOT NULL AND z.continente IS NOT NULL);

QUERY PLAN
HashAggregate (cost=113.34..113.37 rows=2 width=97) (actual time=0.060..0.062 rows=2.00 loops=1)
Group Key: ((z.categoria IS NOT NULL) AND (z.continente IS NOT NULL))
Batches: 1 Memory Usage: 32kB
Buffers: shared hit=3
-> Hash Left Join (cost=63.67..97.64 rows=1570 width=49) (actual time=0.052..0.056 rows=7.00 loops=1)
Hash Cond: (z.id_zona = r_ag.id_zona)
Buffers: shared hit=3
-> Hash Left Join (cost=56.69..86.52 rows=1570 width=52) (actual time=0.016..0.019 rows=7.00 loops=1)
Hash Cond: (z.id_zona = vz_ag.id_zona)
Buffers: shared hit=1


2. Comandos de criação do(s) índice(s)

In [24]:
%%sql zoo
CREATE INDEX idx_vendas_zoo_zona_bid ON vendas_zoo(id_zona, bid);

++
||
++
++

3. Demonstração do custo final (com EXPLAIN ANALYSE)

In [25]:
%%sql zoo
EXPLAIN ANALYSE
SELECT
    (z.categoria IS NOT NULL AND z.continente IS NOT NULL) AS especialidade,
    AVG(vz_ag.receita_zona) AS media_receitas_por_zona,
    AVG(vz_ag.bilhetes_zona) AS media_bilhetes_por_zona,
    AVG(r_ag.votos_zona) AS media_votos_por_zona
FROM zona z
LEFT JOIN(
    SELECT id_zona, SUM(receita) AS receita_zona, COUNT(DISTINCT bid) AS bilhetes_zona
    FROM vendas_zoo
    GROUP BY id_zona
) vz_ag ON z.id_zona = vz_ag.id_zona
LEFT JOIN (
    SELECT id_zona, SUM(votos) AS votos_zona
    FROM recinto
    GROUP BY id_zona
) r_ag ON z.id_zona = r_ag.id_zona
GROUP BY (z.categoria IS NOT NULL AND z.continente IS NOT NULL);

QUERY PLAN
HashAggregate (cost=113.34..113.37 rows=2 width=97) (actual time=0.059..0.062 rows=2.00 loops=1)
Group Key: ((z.categoria IS NOT NULL) AND (z.continente IS NOT NULL))
Batches: 1 Memory Usage: 32kB
Buffers: shared hit=3
-> Hash Left Join (cost=63.67..97.64 rows=1570 width=49) (actual time=0.051..0.055 rows=7.00 loops=1)
Hash Cond: (z.id_zona = r_ag.id_zona)
Buffers: shared hit=3
-> Hash Left Join (cost=56.69..86.52 rows=1570 width=52) (actual time=0.019..0.022 rows=7.00 loops=1)
Hash Cond: (z.id_zona = vz_ag.id_zona)
Buffers: shared hit=1


4. Justificação

Sugerimos o índice idx_vendas_zoo_zona_bid ON vendas_zoo(id_zona, bid) porque a consulta agrega por zona e conta DISTINCT bid. O índice ordena por zona e bilhete, acelerando as agregações e a contagem distinta. Custo: ocupa espaço e pode tornar o refresh da view mais pesado.

#### Entrega

A submissão do projeto deve ser feita na forma de um arquivo zip de nome **entrega-bd-02-GG.zip**, onde **GG** é o número do grupo, estruturado da seguinte forma:

| Ficheiro | Descrição |
| :---- | :---- |
| entrega-bd-02-GG.ipynb | Um ficheiro Jupyter Notebook correspondendo ao preenchimento do template “entrega-bd-02-GG.ipynb” disponibilizado na página da disciplina (onde GG deverá ser subtituído pelo número do grupo).<br/><br/>Deverá preencher a primeira célula com o número do grupo, o número e nome dos alunos que o constituem, tal como a percentagem relativa de contribuição de cada aluno com o respectivo esforço (horas).<br/><br/>Deverá ainda preencher no Jupyter Notebook as respostas a todas as perguntas para as quais há um campo de resposta.<br/><br/>O ficheiro “entrega-bd-02-GG.ipynb” pode ser importado para o ambiente de trabalho disponibilizado para as aulas de laboratório (i.e., https://github.com/bdist/bdist-workspace), que serve de ambiente de teste para as partes em SQL.<br/><br/>Deve certificar-se que todo o código SQL é executável no ambiente de trabalho, |
| app/ | Pasta com os ficheiros que compõem a aplicação web correspondendo à secção<br/><br/>3\. **Desenvolvimento de Aplicação** |

**IMPORTANTE**

* Serão aplicadas penalizações aos grupos que não cumprirem o formato de submissão.
* Não serão aceites submissões fora do prazo.